### ISSUE TITLE
New inlay hint type: implicit conversion hints

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#2254]
It would be useful to draw attention to implicit conversions using an inlay hint that makes them explicit.

I think this could apply to both standard and user-defined conversions.

Example:

```c++
struct UserDefined {
  UserDefined(int x);
};

void foo1(UserDefined);
void foo2(double);

int main() {
  foo1(42);  // inlay hint "(UserDefined)" before the 42
  foo2(43);  // inlay hint "(double)" before the 43
}
```

This would be a new inlay hint kind, so users who don't want them can disable them.

**Comment by 16bit-ykiko** [comment]
Will this conflict with params hint? What about hinting after the expr like “params: 43 as double”?

**Comment by HighCommander4** [comment]
> Will this conflict with params hint?

I don't think so; if both are applicable, it would seem natural to me to put the param hint first, which is what it would look like if a cast was written explicitly into the code.

But now that I think about it a bit more, I realize that using a prefix cast syntax has an ambiguity problem: if there is a binary expression like `a + b` in the code, with no parentheses around it, then we have no way of disambiguating an implicit conversion of the first operand, from an explicit conversion of the binary expression result.

(A suffix `as T` has the same problem, just at the other end.)

This is making me think the hint should actually be "around" the expression, as in `T(expr)`, where `expr` is in the source and `T(` and `)` are hints before and after. It may look a bit odd to have a two-part hint like that; maybe we could omit the parentheses in cases where we don't have an ambiguity?

**Comment by torshepherd** [comment]
Wow, this would be sweet with `as T` spelling. With a quick glance, seems like `ImplicitConversion` nodes are everywhere in the AST though:

![Image](https://github.com/user-attachments/assets/58187fa0-699e-480e-9941-2db5635f406a)

... and sometimes the same portion of text can have multiple ImplicitCasts:

![Image](https://github.com/user-attachments/assets/f160dad7-c5fe-4d41-9091-373b9d6524f7)

But there are tons of crazy hints you could show here. For instance,

```cpp
struct Base {
  void foo() const {}
};
struct Derived : public Base {};

int main() {
  Derived d;
  d.foo();
}
```

This could show two cool things, with multiple spellings:

```cpp
// 1
d as const .foo()
as_const(d).foo()
(const) d.foo()

// 2
d as const Base .foo()
(const Base) d.foo()
```

And for the multiple conversions, how would that be shown?

```cpp
extern void foo(StringRef);

foo("Hello");

// 1
foo(StringRef("Hello"));

// 2
foo(StringRef((const char*) "Hello"));

// 3
foo("Hello" as const char* as StringRef);
```
```
